### PR TITLE
KAFKA-2660; Correct cleanableRatio calculation

### PR DESCRIPTION
**Pull request by lindong28**
@onurkaraman Could you have a look? This is the patch I discussed with you.

### CODE REVIEW

File: core/src/main/scala/kafka/log/LogCleaner.scala [hunk 1]
```diff
@@ -261,9 +261,12 @@ class LogCleaner(val config: CleanerConfig,
  * This class holds the actual logic for cleaning a log
  * @param id An identifier used for logging
  * @param offsetMap The map used for deduplication
- * @param bufferSize The size of the buffers to use. Memory usage will be 2x this number as there is a read and write buffer.
+ * @param ioBufferSize The size of the buffers to use. Memory usage will be 2x this number as there is a read and write buffer.
+ * @param maxIoBufferSize The maximum size of a message that can appear in the log
+ * @param dupBufferLoadFactor The maximum percent full for the deduplication buffer
  * @param throttler The throttler instance to use for limiting I/O rate.
  * @param time The time instance
+ * @param checkDone Check if the cleaning for a partition is aborted. If so, throw an exception
```
Comment (onurkaraman) [inline core/src/main/scala/kafka/log/LogCleaner.scala]: I think the description for the checkDone param is somewhat misleading.

Notice that the Cleaner class uses checkDone but doesn't explicitly throw exceptions. It just so happens that the checkDone function passed into the Cleaner from the CleanerThread uses LogCleanerManager.checkCleaningAborted throws a LogCleaningAbortedException.

File: core/src/main/scala/kafka/log/LogCleaner.scala [hunk 2]
```diff
@@ -673,7 +676,7 @@ private case class CleanerStats(time: Time = SystemTime) {
  * Helper class for a log, its topic/partition, and the last clean position
  */
 private case class LogToClean(topicPartition: TopicAndPartition, log: Log, firstDirtyOffset: Long) extends Ordered[LogToClean] {
-  val cleanBytes = log.logSegments(-1, firstDirtyOffset-1).map(_.size).sum
+  val cleanBytes = log.logSegments(-1, firstDirtyOffset).map(_.size).sum
```
Comment (onurkaraman) [inline core/src/main/scala/kafka/log/LogCleaner.scala:679]: Good catch!

File: core/src/main/scala/kafka/log/LogCleaner.scala (hunk 1 above)
Comment (lindong28) [inline core/src/main/scala/kafka/log/LogCleaner.scala]: Thanks for the review! Fixed now.

**Comment by onurkaraman** [comment]
LGTM

**Comment by lindong28** [comment]
@jjkoshy This is a very small patch and it has been reviewed by Onur. Do you have time to have a look?

**Comment by jjkoshy** [comment]
I agree with the change, but a couple of comments:

I think the way the cleaner works right now, it should be that the dirty offset should always be the start of a segment. So the estimate will be off by at most the size of the last message in the clean section. Can you verify?

Also, it appears from what you described offline that `Log.logSegments` isn't doing what it claims in its doc. I'm not sure that's the case but can you elaborate and if it is an issue consider other usages of the segment range lookup?

Finally, can you add or edit a unit test in `CleanerTest` for this?

**Comment by lindong28** [comment]
@jjkoshy Thanks for your review. Sure, I will provide answer to issues you raised.

**Comment by lindong28** [comment]
@jjkoshy You are right, I have verified that in existing implementation the dirty offset should always be the start of a segment. Therefore the estimate will be off by at most one message size in the clean section.

The documentation of `Log.logSegments(from: Long, to: Long)` would be wrong if either `from` or `to` are is NOT the start of a segment. For example, suppose we have segments with the following offsets: `[1,2,3,4,5], [6,7,8,9], [10,11,12]`, then `Log.logSegments(2, 11)` would only return the segment `[6,7,8]`, even though it should return all 3 segments according to the Doc. In addition, `to` and `from` doesn't seem to always be the start of a segment, e.g. `logSegments(this.recoveryPoint, offset)` in Log.scala.

**Comment by lindong28** [comment]
@jjkoshy I have added unit test to verify the problem and the fix for LogToClean. I also fixed some minor problem that I found in the process of checking log related code. Thanks for taking time to review the patch.

**Comment by lindong28** [comment]
@guozhangwang Would you have time to review this patch?

**Comment by lindong28** [comment]
@gwenshap Would you have time to review this patch? It may be useful to include in 0.9 release since it fixes a bug in log cleaner.

**Comment by gwenshap** [comment]
Super sorry @lindong28, but I just don't know enough about the cleaner to properly review this. 
I hope @guozhangwang, @junrao or @jjkoshy will have time.

**Comment by junrao** [comment]
Thanks for the patch. LGTM

**Comment by lindong28** [comment]
@junrao Thanks for the review! @gwenshap Thanks:)

**Pull request closed without merging**